PEP 1 and 12: Document Topic header

hugovk opened the pull request:
<!--

*Please* read our Contributing Guidelines (CONTRIBUTING.rst)
before submitting an issue or pull request to this repository,
to make sure this repo is the appropriate venue for your proposed change.

Prefix the pull request title with the PEP number; for example:

PEP NNN: Summary of the changes made

-->

Document the Topic header in PEPs 1 and 12.

As suggested by @warsaw in https://github.com/python/peps/pull/2986#issuecomment-1409533600.

The Governance topic will be added in https://github.com/python/peps/pull/2993.

warsaw commented on pep-0001.txt:
```diff
@@ -659,6 +660,9 @@ archives, rather than just a mailto: or hyperlink to the list itself.
 Informational, or Process.
 
+The optional Topic header specifies which topic, if any, the PEP belongs in.
+See https://peps.python.org/topic/ for the existing topics.
```
I don't see `Governance` at this URL.  I might also suggest `peps.python.org/topics` redirect to `/topic` (or vice versa).

CAM-Gerlach replied:
Yeah, I make this mistake all the time. This can be added in the Fastly config here: https://github.com/python/peps/blob/main/infra/main.tf

hugovk replied:
Like this?

```tf
  snippet {
    name    = "topics"
    type    = "recv"
    content = <<-EOT
        if (req.url == "/topics") {
          set req.url = "/topic/";
        }
    EOT
  }
```

Do we also need something for "topics/" ?

CAM-Gerlach replied:
I was about to mention that. If my understanding of the syntax is correct from looking at some of the others, you can do something like ``req.url ~ "^/topics/?$"`` instead to handle that.

warsaw reviewed (approved):
Approved, with comment

CAM-Gerlach commented on pep-0001.txt:
```diff
@@ -659,6 +660,9 @@ archives, rather than just a mailto: or hyperlink to the list itself.
 The Type header specifies the type of PEP: Standards Track,
 Informational, or Process.
 
+The optional Topic header specifies which topic, if any, the PEP belongs in.
```
```suggestion
The optional Topic header lists which special topic(s),
if any, the PEP belongs under.
```

* PEPs can (and very occasionally do) list multiple topics
* Clarify these are _special_ topics, rather than something more general (at least, right now)
* "Belongs under" in this context, not "belongs in"

hugovk replied:
Thanks, I used this but used "topics" instead of the slightly unsightly "topic(s)"

CAM-Gerlach replied:
Yeah, I figured you might not be so fond of that—in that event, given it's such a rare case, I actually think we're better off with the original then. I'll submit another suggestion.

CAM-Gerlach commented on pep-0001.txt:
```diff
@@ -659,6 +660,9 @@ archives, rather than just a mailto: or hyperlink to the list itself.
 Informational, or Process.
 
+The optional Topic header specifies which topic, if any, the PEP belongs in.
+See https://peps.python.org/topic/ for the existing topics.
```
```suggestion
See the :ref:`topic-index` for the existing topics.
```

Use a proper (and named) ref target here, rather than a hardcoded external URL (you'll need to add a ref target label to the topic index page, i.e.:

```rst
.. _topic-index:

Topic Index
***********
```

(You could also use `:doc:` without that, but `:ref:` is preferred as it works regardless of the location of the source or destination file, and won't break if we ever move either).

hugovk replied:
Thanks, done with `:ref:`

CAM-Gerlach reviewed (commented):
It should also be added to the header listing under [How to use this template in PEP 12](https://peps.python.org/pep-0012/#how-to-use-this-template), and maybe also as a short bullet point there.

CAM-Gerlach commented on infra/main.tf:
```diff
@@ -39,6 +39,16 @@ resource "fastly_service_vcl" "peps" {
+    name    = "topics"
+    type    = "recv"
+    content = <<-EOT
+        if (req.url == "/topics") {
+          set req.url = "/topic/";
```
At least judging from the others and skimming the Fastly docs, right now, this not only isn't matching `/`, but also not matching or including URLs that actually specify a topic either, and nor is it performing a redirect (as it needs to set `req.http.location` rather than `req.url`, and then error with the right error code to pass control to the redirect handler.

Following the others, you could do

```suggestion
        if (req.url ~ "^/topics(?:/([\w-]*/?))?$") {
          set req.http.Location = "/topics/" + re.group.1;
          error 618;
```

However, instead of capturing and then replacing, seems like it might be simpler and more robust to just do

```suggestion
        if (req.url ~ "^/topics($|/)") {
          set req.http.Location = regsub(req.http.Location, "^/topics/?", "/topic/");
          error 618;
```

CAM-Gerlach commented:
Oh, and just to be sure, still missing:

> It should also be added to the header listing under [How to use this template in PEP 12](https://peps.python.org/pep-0012/#how-to-use-this-template), and maybe also as a short bullet point there.

hugovk commented:
Thanks, updated!

I also added an `htmlview` target to the `Makefile`, consistent with [CPython docs](https://github.com/python/cpython/blob/0675b8f032c69d265468b31d5cadac6a7ce4bd9c/Doc/Makefile#L138-L140) and [Devguide](https://github.com/python/devguide/blob/21d6d8d4b15c2b355ec170ae7dc523a621951dd3/Makefile#L177-L179). Let me know if this should be in a separate PR. My fingers will be thankful, especially as the docs are output to three different directories: `build``, `build/html`, `_build/html`!

CAM-Gerlach reviewed (approved):
LGTM, thanks @hugovk

warsaw commented on Makefile:
```diff
@@ -11,6 +11,11 @@ RENDER_COMMAND=$(VENVDIR)/bin/python3 build.py -j $(JOBS) -o $(OUTPUT_DIR)
 
+## htmlview       to open the index page built by the html target in your browser
+.PHONY: htmlview
+htmlview: html
```
Or maybe just `view:`?

hugovk replied:
`view` would be quite nice as in `make html view`, but we already have `make htmlview` in at least CPython docs, devguide, pythondotorg, so let's go for consistency.

Thanks!

warsaw reviewed (approved):
Approved, with suggestion
